src/midx.c

```diff
@@ -325,6 +327,46 @@ int git_midx_open(
 	return 0;
 }
 
+bool git_midx_needs_refresh(
+		const git_midx_file *idx,
+		const char *path)
+{
+	git_file fd = -1;
+	struct stat st;
+	ssize_t bytes_read;
+	git_oid idx_checksum = {{0}};
+
+	/* TODO: properly open the file without access time using O_NOATIME */
+	fd = git_futils_open_ro(path);
+	if (fd < 0)
+		return true;
+
+	if (p_fstat(fd, &st) < 0) {
+		p_close(fd);
+		return true;
+	}
+
+	if (!S_ISREG(st.st_mode) ||
+	    !git__is_sizet(st.st_size) ||
+	    (size_t)st.st_size != idx->index_map.len) {
+		p_close(fd);
+		return true;
+	}
+
+	if (p_lseek(fd, -20, SEEK_END) < 0) {
+		p_close(fd);
+		return true;
+	}
+
+	bytes_read = p_read(fd, &idx_checksum, GIT_OID_RAWSZ);
```

Comment:
Are there any cases where `lseek` could back up `-20` from the EOF but then we couldn't read those 20 bytes?  In that case `idx_checksum` might not be fully read and we might be comparing the expected checksum against the uninitialized memory which might give a false positive.

I don't think that this is really possible, I'm mostly just thinking out loud.

One thing that did jump out at me here is that we're using `git_oid`s and `GIT_OID_RAWSZ` but `lseek`ing `-20`.  Maybe we should `lseek` (`0 - GIT_OID_RAWSZ`).  I don't think that we'll ever change that but with the forthcoming sha256 work, I can imagine a world where oids and sizes are a little more complicated, so having that explicitly called out here might be useful?  Then again, there are likely a million places in our codebase where we assume `20`, so...  maybe not worth it until we stumble across it.  Again, mostly just thinking out loud.

Reply:
> Are there any cases where `lseek` could back up `-20` from the EOF but then we couldn't read those 20 bytes? In that case `idx_checksum` might not be fully read and we might be comparing the expected checksum against the uninitialized memory which might give a false positive.

Hmm without the file being modified? I don't think so, unless there's a freak coincidence where the last 20 bytes being in an unreadable sector or a corrupted block or something similarly implausible. the good thing is that the memory is initialized to 0 in L337! but it may be a good idea to change the condition in L364 to be `!= GIT_OID_RAWSZ`.

probably the "correct" thing to do would be to introduce support for [`pread(2)`](https://man7.org/linux/man-pages/man2/pread.2.html) (#5583 comes to mind) to avoid this seek-then-read shenanigans.

> I don't think that this is really possible, I'm mostly just thinking out loud.
> 
> One thing that did jump out at me here is that we're using `git_oid`s and `GIT_OID_RAWSZ` but `lseek`ing `-20`. Maybe we should `lseek` (`0 - GIT_OID_RAWSZ`). I don't think that we'll ever change that but with the forthcoming sha256 work, I can imagine a world where oids and sizes are a little more complicated, so having that explicitly called out here might be useful? Then again, there are likely a million places in our codebase where we assume `20`, so... maybe not worth it until we stumble across it. Again, mostly just thinking out loud.

ah that's a good idea!